services/ndd_service.js

```diff
@@ -4,155 +4,106 @@
-      if (inspectorUrl.startsWith('http:')) {
-        const targetInfo = (await this._fetch(inspectorUrl))[0];
-        inspectorUrl = targetInfo.webSocketDebuggerUrl;
-      }
-      const ws = new WebSocket(inspectorUrl);
-      ws.once('open', () => {
-        this._sockets.set(id, ws);
-        this._frontend.detected(id, info);
-      });
-      ws.on('message', rawMessage => {
-        const message = JSON.parse(rawMessage);
-        message.sessionId = id;
-        this._frontend.dispatchMessage(message);
+    const localRelative = path.relative(process.cwd(), `${os.tmpdir()}/`);
+    const pipePrefix = process.platform === 'win32' ? '\\\\.\\pipe\\' : localRelative;
+    const pipeName = `node-ndb.${process.pid}.sock`;
+    this._pipe = path.join(pipePrefix, pipeName);
+    const server = net.createServer(socket => {
+      socket.on('data', async d => {
+        await new Promise(resolve => this._runSession(JSON.parse(d.toString()), frontend, resolve));
+        socket.write('ack');
       });
-      ws.once('close', () => {
-        this._sockets.delete(id);
-        this._frontend.disconnected(id);
-      });
-      ws.once('error', () => 0);
-    } catch (e) {
-    }
+      socket.on('error', e => caughtErrorDebug(e));
+    }).listen(this._pipe);
+    server.unref();
   }
 
-  _fetch(url) {
-    return new Promise(resolve => {
-      http.get(url, res => {
-        if (res.statusCode !== 200) {
-          res.resume();
-          resolve(null);
-        } else {
-          res.setEncoding('utf8');
-          let buffer = '';
-          res.on('data', data => buffer += data);
-          res.on('end', _ => {
-            try {
-              resolve(JSON.parse(buffer));
-            } catch (e) {
-              resolve(null);
-            }
-          });
-        }
-      }).on('error', _ => resolve(null));
-    });
+  dispose() {
+    process.disconnect();
   }
 
-  async dispose() {
-    try {
-      for (const id of Array.from(this._running)) {
-        try {
-          process.kill(id, 'SIGKILL');
-        } catch (e) {
-        }
-      }
-      this._running.clear();
-      for (const ws of this._sockets.values())
-        ws.close();
-      for (const watcher of this._nddStoreWatchers)
-        watcher.close();
-      await removeFolder(this._nddStores[0]);
-      this._nddStores = [];
-      this._nddStoreWatchers = [];
-    } catch (e) {
-    } finally {
-      process.exit(0);
+  async _runSession(info, frontend, resumeTarget) {
+    const ws = new WebSocket(info.inspectorUrl);
+    const openPromise = new Promise(resolve => ws.once('open', () => resolve(DebugState.WS_OPEN)));
+    const errorPromise = new Promise(resolve => ws.once('error', () => resolve(DebugState.WS_ERROR)));
+    const closePromise = new Promise(resolve => ws.once('close', () => resolve(DebugState.WS_CLOSE)));
+    let state = await Promise.race([openPromise, errorPromise, closePromise, this._disconnectPromise]);
+    if (state === DebugState.WS_OPEN) {
```

Comment:
I think this code can be simplified by using `openPromise.then`

Reply:
disconnect might happen in the middle of ws open, I do not need to connect in this case.